parquet/src/main/java/org/apache/iceberg/parquet/ParquetMetrics.java

```diff
@@ -0,0 +1,613 @@
+        nullValueCounts.put(id, metrics.nullValueCount());
+      }
+
+      if (metrics.nanValueCount() >= 0) {
+        nanValueCounts.put(id, metrics.nanValueCount());
+      }
+
+      if (metrics.hasBounds()) {
+        lowerBounds.put(id, metrics.lowerBound());
+        upperBounds.put(id, metrics.upperBound());
+      }
+    }
+
+    return new Metrics(
+        rowCount,
+        columnSizes,
+        valueCounts,
+        nullValueCounts,
+        nanValueCounts,
+        lowerBounds,
+        upperBounds);
+  }
+
+  private static class MetricsVisitor
+      extends TypeWithSchemaVisitor<Iterable<FieldMetrics<ByteBuffer>>> {
+    private final Schema schema;
+    private final MetricsConfig metricsConfig;
+    private final Map<Integer, FieldMetrics<?>> metricsById;
+    private final Multimap<ColumnPath, ColumnChunkMetaData> columns;
+
+    private MetricsVisitor(
+        Schema schema,
+        MetricsConfig metricsConfig,
+        Map<Integer, FieldMetrics<?>> metricsById,
+        Multimap<ColumnPath, ColumnChunkMetaData> columns) {
+      this.schema = schema;
+      this.metricsConfig = metricsConfig;
+      this.metricsById = metricsById;
+      this.columns = columns;
+    }
+
+    @Override
+    public Iterable<FieldMetrics<ByteBuffer>> message(
+        Types.StructType iStruct,
+        MessageType message,
+        List<Iterable<FieldMetrics<ByteBuffer>>> fieldResults) {
+      return Iterables.concat(fieldResults);
+    }
+
+    @Override
+    public Iterable<FieldMetrics<ByteBuffer>> struct(
+        Types.StructType iStruct,
+        GroupType struct,
+        List<Iterable<FieldMetrics<ByteBuffer>>> fieldResults) {
+      return Iterables.concat(fieldResults);
+    }
+
+    @Override
+    public Iterable<FieldMetrics<ByteBuffer>> list(
+        Types.ListType iList, GroupType array, Iterable<FieldMetrics<ByteBuffer>> elementResults) {
+      // remove lower and upper bounds for repeated fields
+      return ImmutableList.of();
+    }
+
+    @Override
+    public Iterable<FieldMetrics<ByteBuffer>> map(
+        Types.MapType iMap,
+        GroupType map,
+        Iterable<FieldMetrics<ByteBuffer>> keyResults,
+        Iterable<FieldMetrics<ByteBuffer>> valueResults) {
+      // repeated fields are not currently supported
+      return ImmutableList.of();
+    }
+
+    @Override
+    public Iterable<FieldMetrics<ByteBuffer>> primitive(
+        org.apache.iceberg.types.Type.PrimitiveType iPrimitive, PrimitiveType primitive) {
+      Type.ID id = primitive.getId();
+      if (null == id) {
+        return ImmutableList.of();
+      }
+      int fieldId = id.intValue();
+
+      MetricsModes.MetricsMode mode = MetricsUtil.metricsMode(schema, metricsConfig, fieldId);
+      if (mode == MetricsModes.None.get()) {
+        return ImmutableList.of();
+      }
+
+      int length = truncateLength(mode);
+
+      FieldMetrics<ByteBuffer> metrics = metricsFromFieldMetrics(fieldId, iPrimitive, length);
```

Comment:
The main value of doing this in the visitor is to avoid processing the column metadata if there is already a `FieldMetrics` object available.